components/ai_chat/renderer/page_text_distilling.cc

```diff
@@ -204,4 +225,102 @@ void DistillPageText(
+  /**
+   * Concatenates our injector-script (retrieved from the resource bundle) with
+   * our extractor script. Because the injector-script is wrapped in an IIFE, we
+   * leverage custom events dispatched against the window object to communicate
+   * between the two scripts.
+   *
+   * The extractor script is currently hard-coded to request content at level 3.
+   * Level 3 is the "FULL" level, which is the highest content level. The level
+   * could be tied to the currently-active model (e.g., smaller models could be
+   * defaulted to a smaller level).
+   */
+  std::string script = script_content + R"(
+    new Promise(function(resolve, reject) {
+        window.addEventListener('{complete_key}', (event) => {
+            if (event?.detail?.result) {
+                return resolve(event.detail.result);
+            }
+            reject("No result in {complete_key} event");
+        }, { once: true });
+        window.dispatchEvent(new CustomEvent(
+          '{initiate_key}', { detail: { level: 3 } }
+        ));
+    });
+  )";
+
+  base::ReplaceSubstringsAfterOffset(&script, 0, "{initiate_key}",
+                                     "LEO_DISTILL_REQUESTED");
+  base::ReplaceSubstringsAfterOffset(&script, 0, "{complete_key}",
+                                     "LEO_DISTILL_RESULT");
+
+  blink::WebScriptSource source =
+      blink::WebScriptSource(blink::WebString::FromUTF8(script));
+
+  auto on_script_executed =
+      [](base::OnceCallback<void(const std::optional<std::string>&)> callback,
+         std::optional<base::Value> value, base::TimeTicks start_time) {
+        if (value && value->is_string() && !value->GetString().empty()) {
+          std::move(callback).Run(value->GetString());
+          return;
+        }
+
+        std::move(callback).Run({});
+      };
+
+  // Execute the combined script as a single source
+  render_frame->GetWebFrame()->RequestExecuteScript(
+      world_id, base::span_from_ref(source),
+      blink::mojom::UserActivationOption::kDoNotActivate,
+      blink::mojom::EvaluationTiming::kAsynchronous,
+      blink::mojom::LoadEventBlockingOption::kDoNotBlock,
+      base::BindOnce(on_script_executed, std::move(callback)),
+      blink::BackForwardCacheAware::kAllow,
+      blink::mojom::WantResultOption::kWantResult,
+      /**
+       * Because we are using a promise to resolve the result, we will
+       * use the `kAwait` option to ensure the promise is resolved before
+       * the callback is invoked.
+       */
+      blink::mojom::PromiseResultOption::kAwait);
+}
+
+bool LoadSiteScriptForHost(std::string* host,
+                           std::string* script_content,
+                           bool* needs_main_world) {
+  std::string lower_host = base::ToLowerASCII(*host);
+
+  if (base::StartsWith(lower_host, "www.",
+                       base::CompareCase::INSENSITIVE_ASCII)) {
+    lower_host = lower_host.substr(4);
+  }
+
+  struct ScriptResource {
```

Comment:
you generally shouldn't have a struct definition inside a method body, but I think maybe this is making things overly complicated given that all of this is going to have to change to handle delivering these scripts by component. I would just use `std::pair` here and also return `std::optional<std::pair<std::string_view, bool>>`